File: clients/typescript/src/protocol.ts
/**
 * Binary wire protocol: command/status codes and the encode/decode of payload
 * bodies, with no I/O. Mirrors the Rust definitions in
 * crates/zydecodb-engine/src/frame.rs and crates/zydecodb-document/src/wire.rs
 * and is verified byte-for-byte against clients/conformance/vectors.json.
 */

export const PROTO_VERSION = 0x01;
export const HEADER_LEN = 6;

// Command codes (envelope byte 1).
export const Cmd = {
  Put: 0x01,
  Get: 0x02,
  Del: 0x03,
  Begin: 0x10,
  Commit: 0x11,
  Rollback: 0x12,
  Query: 0x20,
  DocPut: 0x21,
  DocDel: 0x22,
  Find: 0x23,
  Update: 0x24,
  Delete: 0x25,
  Count: 0x26,
  DocGetRev: 0x27,
  FindRev: 0x28,
  DocPutIfMatch: 0x29,
  DocUpdateIfMatch: 0x2a,
  Aggregate: 0x2b,
  Watch: 0x2c,
  IndexDef: 0x30,
  SessionInit: 0x40,
  Ping: 0xf0,
  Stats: 0xf1,
} as const;

// Query / count sub-command discriminators (first payload byte).
const QUERY_BY_ID = 0x00;
const QUERY_INDEX_RANGE = 0x01;
const COUNT_MODE_COUNT = 0x00;
const COUNT_MODE_DISTINCT = 0x01;

// Projection modes for Find.
export const Proj = {
  None: 0x00,
  Include: 0x01,
  Exclude: 0x02,
} as const;

// Bits of the optional trailing flags byte on write payloads.
const FLAG_RELAXED = 0x01;
const FLAG_UPSERT = 0x02;

const WATCH_FRAME_ACK = 0x01;
const WATCH_FRAME_EVENT = 0x02;
const WATCH_FRAME_HEARTBEAT = 0x03;

export const WatchOp = {
  Upsert: 0x01,
  Delete: 0x02,
} as const;

// Status codes (response envelope byte 1).
export const Status = {
  Ok: 0x00,
  NotFound: 0x01,
  Error: 0x02,
  Conflict: 0x03,
  IoError: 0x04,
  InvalidKey: 0x05,
  InvalidValue: 0x06,
  EngineBusy: 0x07,
  ProtocolError: 0x08,
  PolicyRejected: 0x09,
  UnsupportedFormat: 0x0a,
  Unauthorized: 0x0b,
  Forbidden: 0x0c,
} as const;

const STATUS_NAMES: Record<number, string> = {
  [Status.Ok]: "Ok",
  [Status.NotFound]: "NotFound",
  [Status.Error]: "Error",
  [Status.Conflict]: "Conflict",
  [Status.IoError]: "IoError",
  [Status.InvalidKey]: "InvalidKey",
  [Status.InvalidValue]: "InvalidValue",
  [Status.EngineBusy]: "EngineBusy",
  [Status.ProtocolError]: "ProtocolError",
  [Status.PolicyRejected]: "PolicyRejected",
  [Status.UnsupportedFormat]: "UnsupportedFormat",
  [Status.Unauthorized]: "Unauthorized",
  [Status.Forbidden]: "Forbidden",
};

export function statusName(status: number): string {
  return STATUS_NAMES[status] ?? `0x${status.toString(16).padStart(2, "0")}`;
}

/** Build the 6-byte request envelope header. */
export function encodeHeader(command: number, payloadLen: number): Buffer {
  const h = Buffer.allocUnsafe(HEADER_LEN);
  h.writeUInt8(PROTO_VERSION, 0);
  h.writeUInt8(command, 1);
  h.writeUInt32BE(payloadLen >>> 0, 2);
  return h;
}

/** Build a Put payload. */
export function encodePut(key: Buffer, value: Buffer, expiresAt: number | bigint = 0): Buffer {
  const buf = Buffer.allocUnsafe(16 + 8 + 8 + 4 + 4 + key.length + value.length);
  buf.fill(0, 0, 16); // routing key
  buf.writeBigUInt64BE(0n, 16); // txid
  buf.writeBigUInt64BE(BigInt(expiresAt), 24);
  buf.writeUInt32BE(key.length, 32);
  buf.writeUInt32BE(value.length, 36);
  key.copy(buf, 40);
  value.copy(buf, 40 + key.length);
  return buf;
}

/** Build a Key payload for Get/Del. */
export function encodeKey(key: Buffer): Buffer {
  const buf = Buffer.allocUnsafe(16 + 8 + 4 + key.length);
  buf.fill(0, 0, 16); // routing key
  buf.writeBigUInt64BE(0n, 16); // snapshot_seq
  buf.writeUInt32BE(key.length, 24);
  key.copy(buf, 28);
  return buf;
}

function lp(b: Buffer): Buffer {
  const len = Buffer.allocUnsafe(4);
  len.writeUInt32BE(b.length >>> 0, 0);
  return Buffer.concat([len, b]);
}

function u32(value: number): Buffer {
  const b = Buffer.allocUnsafe(4);
  b.writeUInt32BE(value >>> 0, 0);
  return b;
}

function flag(relaxed: boolean, upsert = false): Buffer {
  let flags = 0;
  if (relaxed) flags |= FLAG_RELAXED;
  if (upsert) flags |= FLAG_UPSERT;
  return Buffer.from([flags]);
}

function boolByte(v: boolean): Buffer {
  return Buffer.from([v ? 1 : 0]);
}

function utf8(s: string): Buffer {
  return Buffer.from(s, "utf8");
}

/**
 * DocPut payload: [collection][doc_id][body][flags][optional expires_at u64 BE].
 * `body` is already-serialized JSON (the codec treats it as opaque bytes).
 * `expiresAt` is absolute unix millis; 0 omits the trailer (never expires).
 */
export function encodeDocPut(
  collection: string,
  docId: Buffer,
  body: Buffer,
  relaxed: boolean,
  expiresAt: number | bigint = 0,
): Buffer {
  const parts = [lp(utf8(collection)), lp(docId), lp(body), flag(relaxed)];
  const exp = BigInt(expiresAt);
  if (exp !== 0n) {
    const e = Buffer.allocUnsafe(8);
    e.writeBigUInt64BE(exp, 0);
    parts.push(e);
  }
  return Buffer.concat(parts);
}

/** Conditional replace: [collection][doc_id][body][flags][if_match u64][optional expires_at]. */
export function encodeDocPutIfMatch(
  collection: string,
  docId: Buffer,
  body: Buffer,
  relaxed: boolean,
  ifMatch: bigint | number,
  expiresAt: number | bigint = 0,
): Buffer {
  const m = Buffer.allocUnsafe(8);
  m.writeBigUInt64BE(BigInt(ifMatch), 0);
  const parts = [lp(utf8(collection)), lp(docId), lp(body), flag(relaxed), m];
  const exp = BigInt(expiresAt);
  if (exp !== 0n) {
    const e = Buffer.allocUnsafe(8);
    e.writeBigUInt64BE(exp, 0);
    parts.push(e);
  }
  return Buffer.concat(parts);
}

/** Conditional by-id update: [collection][doc_id][update][flags][if_match u64]. */
export function encodeDocUpdateIfMatch(
  collection: string,
  docId: Buffer,
  update: Buffer,
  relaxed: boolean,
  ifMatch: bigint | number,
): Buffer {
  const m = Buffer.allocUnsafe(8);
  m.writeBigUInt64BE(BigInt(ifMatch), 0);
  return Buffer.concat([lp(utf8(collection)), lp(docId), lp(update), flag(relaxed), m]);
}

/** DocDel payload: [collection][doc_id]. */
export function encodeDocDel(collection: string, docId: Buffer): Buffer {
  return Buffer.concat([lp(utf8(collection)), lp(docId)]);
}

/** Trailer tag for per-field IndexDef directions (1=ASC, 0=DESC). */
export const INDEX_DIR_TAG = 0x02;

/** IndexDef payload: [collection][index][unique u8][field_count u32]{[field]}
 * [optional expire_after_seconds u64][optional 0x02 + N direction bytes]. */
export function encodeIndexDef(
  collection: string,
  index: string,
  fields: string[],
  unique: boolean,
  expireAfterSeconds: number | bigint = 0,
  directions?: boolean[],
): Buffer {
  const parts = [lp(utf8(collection)), lp(utf8(index)), boolByte(unique), u32(fields.length)];
  for (const f of fields) parts.push(lp(utf8(f)));
  const dirs = directions ?? [];
  const anyDesc =
    dirs.length === fields.length && fields.length > 0 && dirs.some((d) => !d);
  const exp = BigInt(expireAfterSeconds);
  if (exp !== 0n || anyDesc) {
    const e = Buffer.allocUnsafe(8);
    e.writeBigUInt64BE(exp, 0);
    parts.push(e);
  }
  if (anyDesc) {
    parts.push(Buffer.from([INDEX_DIR_TAG]));
    parts.push(Buffer.from(dirs.map((d) => (d ? 1 : 0))));
  }
  return Buffer.concat(parts);
}

/** By-id Query payload: [mode][collection][doc_id]. */
export function encodeQueryById(collection: string, docId: Buffer): Buffer {
  return Buffer.concat([Buffer.from([QUERY_BY_ID]), lp(utf8(collection)), lp(docId)]);
}

/**
 * Index-range Query payload: [mode][collection][index][limit u32][lo][hi][cursor].
 * `lo`/`hi` are JSON-array bound bytes (empty = unbounded); `cursor` is an
 * opaque page token (empty = first page).
 */
export function encodeQueryIndexRange(
  collection: string,
  index: string,
  lo: Buffer,
  hi: Buffer,
  cursor: Buffer,
  limit: number,
): Buffer {
  return Buffer.concat([
    Buffer.from([QUERY_INDEX_RANGE]),
    lp(utf8(collection)),
    lp(utf8(index)),
    u32(limit),
    lp(lo),
    lp(hi),
    lp(cursor),
  ]);
}

/** One ordering term: a dotted field path and its direction. */
export interface SortKey {
  field: string;
  ascending: boolean;
}

/** Field selection. `mode` is Proj.None/Include/Exclude. */
export interface Projection {
  mode: number;
  fields: string[];
}

/** Find payload. `filter` is opaque JSON bytes (empty = match all). */
export function encodeFind(
  collection: string,
  filter: Buffer,
  sort: SortKey[],
  projection: Projection,
  skip: number,
  limit: number,
  cursor: Buffer,
): Buffer {
  const parts = [lp(utf8(collection)), lp(filter), u32(sort.length)];
  for (const s of sort) {
    parts.push(lp(utf8(s.field)), boolByte(s.ascending));
  }
  if (projection.mode === Proj.Include || projection.mode === Proj.Exclude) {
    parts.push(Buffer.from([projection.mode]), u32(projection.fields.length));
    for (const f of projection.fields) parts.push(lp(utf8(f)));
  } else {
    parts.push(Buffer.from([Proj.None]));
  }
  parts.push(u32(skip), u32(limit), lp(cursor));
  return Buffer.concat(parts);
}

/** Update payload: [collection][filter][update][multi u8][flags]. */
export function encodeUpdate(
  collection: string,
  filter: Buffer,
  update: Buffer,
  multi: boolean,
  relaxed: boolean,
  upsert = false,
): Buffer {
  return Buffer.concat([
    lp(utf8(collection)),
    lp(filter),
    lp(update),
    boolByte(multi),
    flag(relaxed, upsert),
  ]);
}

/** Filter-based Delete payload: [collection][filter][multi u8][flags]. */
export function encodeDelete(
  collection: string,
  filter: Buffer,
  multi: boolean,
  relaxed: boolean,
): Buffer {
  return Buffer.concat([lp(utf8(collection)), lp(filter), boolByte(multi), flag(relaxed)]);
}

/** Count payload: [mode][collection][filter]. */
export function encodeCount(collection: string, filter: Buffer): Buffer {
  return Buffer.concat([Buffer.from([COUNT_MODE_COUNT]), lp(utf8(collection)), lp(filter)]);
}

/** Distinct payload: [mode][collection][filter][field]. */
export function encodeDistinct(collection: string, filter: Buffer, field: string): Buffer {
  return Buffer.concat([
    Buffer.from([COUNT_MODE_DISTINCT]),
    lp(utf8(collection)),
    lp(filter),
    lp(utf8(field)),
  ]);
}

/** Aggregate payload: [collection][pipeline_json]. */
export function encodeAggregate(collection: string, pipeline: Buffer): Buffer {
  return Buffer.concat([lp(utf8(collection)), lp(pipeline)]);
}

/** Watch request: [collection lp][resume_token lp]. */
export function encodeWatch(collection: string, resumeToken: Buffer = Buffer.alloc(0)): Buffer {
  return Buffer.concat([lp(utf8(collection)), lp(resumeToken)]);
}

export interface WatchFrame {
  kind: "ack" | "event" | "heartbeat";
  resumeToken: Buffer;
  op?: number;
  docId?: Buffer;
  body?: Buffer;
}

/** Decode one Watch stream frame from an Ok response payload. */
export function decodeWatchFrame(buf: Buffer): WatchFrame {
  if (buf.length === 0) throw new Error("zydecodb: empty watch frame");
  let off = 0;
  const kindByte = buf.readUInt8(off);
  off += 1;
  const need = (n: number): void => {
    if (off + n > buf.length) throw new Error("zydecodb: payload truncated");
  };
  need(4);
  const tlen = buf.readUInt32BE(off);
  off += 4;
  need(tlen);
  const resumeToken = buf.subarray(off, off + tlen);
  off += tlen;
  if (kindByte === WATCH_FRAME_ACK) {
    return { kind: "ack", resumeToken: Buffer.from(resumeToken) };
  }
  if (kindByte === WATCH_FRAME_HEARTBEAT) {
    return { kind: "heartbeat", resumeToken: Buffer.from(resumeToken) };
  }
  if (kindByte === WATCH_FRAME_EVENT) {
    need(1);
    const op = buf.readUInt8(off);
    off += 1;
    need(4);
    const idlen = buf.readUInt32BE(off);
    off += 4;
    need(idlen);
    const docId = buf.subarray(off, off + idlen);
    off += idlen;
    need(4);
    const blen = buf.readUInt32BE(off);
    off += 4;
    need(blen);
    const body = buf.subarray(off, off + blen);
    return {
      kind: "event",
      resumeToken: Buffer.from(resumeToken),
      op,
      docId: Buffer.from(docId),
      body: Buffer.from(body),
    };
  }
  throw new Error(`zydecodb: unknown watch frame 0x${kindByte.toString(16).padStart(2, "0")}`);
}

/** Decode Aggregate response: [row_count u32]{[row_json lp]}*. */
export function decodeAggregateResponse(buf: Buffer): Buffer[] {
  let off = 0;
  const need = (n: number): void => {
    if (off + n > buf.length) throw new Error("zydecodb: payload truncated");
  };
  need(4);
  const count = buf.readUInt32BE(off);
  off += 4;
  const rows: Buffer[] = [];
  for (let i = 0; i < count; i++) {
    need(4);
    const n = buf.readUInt32BE(off);
    off += 4;
    need(n);
    rows.push(buf.subarray(off, off + n));
    off += n;
  }
  return rows;
}

/** One decoded row from a query/find response page. */
export interface Row {
  docId: Buffer;
  body: Buffer;
  /** Present only when decoded via {@link decodePageWithRevision}. */
  revision?: bigint;
}

/** A decoded response page. `cursor` is null when there are no more pages. */
export interface Page {
  rows: Row[];
  cursor: Buffer | null;
}

/** Decode a response page: [row_count u32]{[doc_id][body]}[cursor]. */
export function decodePage(buf: Buffer): Page {
  return decodePageInner(buf, false);
}

/** Decode a FindRev page (each row appends an 8-byte revision). */
export function decodePageWithRevision(buf: Buffer): Page {
  return decodePageInner(buf, true);
}

function decodePageInner(buf: Buffer, withRevision: boolean): Page {
  let off = 0;
  const need = (n: number): void => {
    if (off + n > buf.length) throw new Error("zydecodb: payload truncated");
  };
  need(4);
  const count = buf.readUInt32BE(off);
  off += 4;
  const rows: Row[] = [];
  for (let i = 0; i < count; i++) {
    need(4);
    const klen = buf.readUInt32BE(off);
    off += 4;
    need(klen);
    const docId = buf.subarray(off, off + klen);
    off += klen;
    need(4);
    const blen = buf.readUInt32BE(off);
    off += 4;
    need(blen);
    const body = buf.subarray(off, off + blen);
    off += blen;
    let revision: bigint | undefined;
    if (withRevision) {
      need(8);
      revision = buf.readBigUInt64BE(off);
      off += 8;
    }
    rows.push({ docId, body, revision });
  }
  need(4);
  const clen = buf.readUInt32BE(off);
  off += 4;
  need(clen);
  const cursorBytes = buf.subarray(off, off + clen);
  return { rows, cursor: cursorBytes.length === 0 ? null : cursorBytes };
}

/** Decode a DocGetRev response: [body lp][revision u64 BE]. */
export function decodeDocGetRevResponse(buf: Buffer): { body: Buffer; revision: bigint } {
  if (buf.length < 4) throw new Error("zydecodb: payload truncated");
  const blen = buf.readUInt32BE(0);
  if (buf.length < 4 + blen + 8) throw new Error("zydecodb: payload truncated");
  const body = buf.subarray(4, 4 + blen);
  const revision = buf.readBigUInt64BE(4 + blen);
  return { body, revision };
}

/** Decode Begin response: [tx_id u64 BE][snapshot_seq u64 BE]. */
export function decodeBeginResponse(buf: Buffer): { txId: bigint; snapshotSeq: bigint } {
  if (buf.length !== 16) throw new Error("zydecodb: Begin response must be 16 bytes");
  return { txId: buf.readBigUInt64BE(0), snapshotSeq: buf.readBigUInt64BE(8) };
}

/** Decode Commit response: [seq u64 BE]. */
export function decodeCommitResponse(buf: Buffer): bigint {
  if (buf.length !== 8) throw new Error("zydecodb: Commit response must be 8 bytes");
  return buf.readBigUInt64BE(0);
}

/** Decode stage ack: [logical_ops u32 BE][estimated_keys u32 BE]. */
export function decodeStageAck(buf: Buffer): { logicalOps: number; estimatedKeys: number } {
  if (buf.length !== 8) throw new Error("zydecodb: stage ack must be 8 bytes");
  return { logicalOps: buf.readUInt32BE(0), estimatedKeys: buf.readUInt32BE(4) };
}

File: clients/python/zydecodb/_protocol.py
"""Binary wire protocol: command codes, status codes, and payload codecs.

This module is pure encoding/decoding with no I/O, so it can be unit-tested in
isolation and reused by the connection layer. It mirrors the Rust definitions in
`crates/zydecodb-engine/src/frame.rs` and `crates/zydecodb-document/src/wire.rs`.
"""

from __future__ import annotations

import json
import struct
from typing import Any, List, Optional, Tuple

PROTO_VERSION = 0x01
HEADER_LEN = 6

# --- commands ---
CMD_PUT = 0x01
CMD_GET = 0x02
CMD_DEL = 0x03
CMD_BEGIN = 0x10
CMD_COMMIT = 0x11
CMD_ROLLBACK = 0x12
CMD_QUERY = 0x20
CMD_DOC_PUT = 0x21
CMD_DOC_DEL = 0x22
CMD_FIND = 0x23
CMD_UPDATE = 0x24
CMD_DELETE = 0x25
CMD_COUNT = 0x26
CMD_DOC_GET_REV = 0x27
CMD_FIND_REV = 0x28
CMD_DOC_PUT_IF_MATCH = 0x29
CMD_DOC_UPDATE_IF_MATCH = 0x2A
CMD_AGGREGATE = 0x2B
CMD_WATCH = 0x2C
CMD_INDEX_DEF = 0x30
CMD_SESSION_INIT = 0x40
CMD_PING = 0xF0
CMD_STATS = 0xF1

# --- query / count sub-commands ---
QUERY_BY_ID = 0x00
QUERY_INDEX_RANGE = 0x01
COUNT_MODE_COUNT = 0x00
COUNT_MODE_DISTINCT = 0x01

# --- projection modes ---
PROJ_NONE = 0x00
PROJ_INCLUDE = 0x01
PROJ_EXCLUDE = 0x02

# Bit 0 of the optional trailing flags byte on write payloads.
FLAG_RELAXED = 0x01
FLAG_UPSERT = 0x02

# --- watch stream frame kinds (first byte of Ok payloads) ---
WATCH_FRAME_ACK = 0x01
WATCH_FRAME_EVENT = 0x02
WATCH_FRAME_HEARTBEAT = 0x03

WATCH_OP_UPSERT = 0x01
WATCH_OP_DELETE = 0x02
STATUS_OK = 0x00
STATUS_NOT_FOUND = 0x01
STATUS_ERROR = 0x02
STATUS_CONFLICT = 0x03
STATUS_IO_ERROR = 0x04
STATUS_INVALID_KEY = 0x05
STATUS_INVALID_VALUE = 0x06
STATUS_ENGINE_BUSY = 0x07
STATUS_PROTOCOL_ERROR = 0x08
STATUS_POLICY_REJECTED = 0x09
STATUS_UNSUPPORTED_FORMAT = 0x0A
STATUS_UNAUTHORIZED = 0x0B
STATUS_FORBIDDEN = 0x0C

STATUS_NAMES = {
    STATUS_OK: "Ok",
    STATUS_NOT_FOUND: "NotFound",
    STATUS_ERROR: "Error",
    STATUS_CONFLICT: "Conflict",
    STATUS_IO_ERROR: "IoError",
    STATUS_INVALID_KEY: "InvalidKey",
    STATUS_INVALID_VALUE: "InvalidValue",
    STATUS_ENGINE_BUSY: "EngineBusy",
    STATUS_PROTOCOL_ERROR: "ProtocolError",
    STATUS_POLICY_REJECTED: "PolicyRejected",
    STATUS_UNSUPPORTED_FORMAT: "UnsupportedFormat",
    STATUS_UNAUTHORIZED: "Unauthorized",
    STATUS_FORBIDDEN: "Forbidden",
}


def encode_header(command: int, payload_len: int) -> bytes:
    return struct.pack(">BBI", PROTO_VERSION, command, payload_len)


def _lp(b: bytes) -> bytes:
    """Length-prefix a byte string with a u32 big-endian length."""
    return struct.pack(">I", len(b)) + b


def _json_bytes(value: Any) -> bytes:
    return json.dumps(value, separators=(",", ":")).encode("utf-8")


def encode_put(
    key: bytes,
    value: bytes,
    *,
    routing_key: bytes = b"\x00" * 16,
    txid: int = 0,
    expires_at: int = 0,
) -> bytes:
    return (
        routing_key[:16].ljust(16, b"\x00")
        + struct.pack(">QQII", txid, expires_at, len(key), len(value))
        + key
        + value
    )


def encode_key(
    key: bytes,
    *,
    routing_key: bytes = b"\x00" * 16,
    snapshot_seq: int = 0,
) -> bytes:
    return (
        routing_key[:16].ljust(16, b"\x00")
        + struct.pack(">QI", snapshot_seq, len(key))
        + key
    )


INDEX_DIR_TAG = 0x02


def encode_index_def(
    collection: str,
    index: str,
    fields: List[str],
    *,
    unique: bool,
    expire_after_seconds: int = 0,
    directions: Optional[List[bool]] = None,
) -> bytes:
    out = _lp(collection.encode()) + _lp(index.encode())
    out += bytes([1 if unique else 0])
    out += struct.pack(">I", len(fields))
    for field in fields:
        out += _lp(field.encode())
    dirs = list(directions) if directions is not None else []
    any_desc = len(dirs) == len(fields) and any(not d for d in dirs)
    if expire_after_seconds or any_desc:
        out += struct.pack(">Q", int(expire_after_seconds))
    if any_desc:
        out += bytes([INDEX_DIR_TAG])
        out += bytes([1 if d else 0 for d in dirs])
    return out


def encode_doc_put(
    collection: str,
    doc_id: str,
    document: Any,
    *,
    relaxed: bool,
    expires_at: int = 0,
) -> bytes:
    """DocPut payload: [collection][doc_id][body][flags][optional expires_at u64 BE].

    ``expires_at`` is absolute unix millis; ``0`` omits the trailer (never expires).
    """
    out = _lp(collection.encode()) + _lp(doc_id.encode()) + _lp(_json_bytes(document))
    out += bytes([FLAG_RELAXED if relaxed else 0])
    if expires_at:
        out += struct.pack(">Q", int(expires_at))
    return out


def encode_doc_put_if_match(
    collection: str,
    doc_id: str,
    document: Any,
    *,
    relaxed: bool,
    if_match: int,
    expires_at: int = 0,
) -> bytes:
    """Conditional replace: [collection][doc_id][body][flags][if_match][optional expires_at]."""
    out = _lp(collection.encode()) + _lp(doc_id.encode()) + _lp(_json_bytes(document))
    out += bytes([FLAG_RELAXED if relaxed else 0])
    out += struct.pack(">Q", int(if_match))
    if expires_at:
        out += struct.pack(">Q", int(expires_at))
    return out


def encode_doc_update_if_match(
    collection: str,
    doc_id: str,
    update_doc: Any,
    *,
    relaxed: bool,
    if_match: int,
) -> bytes:
    """Conditional by-id update: [collection][doc_id][update][flags][if_match]."""
    return (
        _lp(collection.encode())
        + _lp(doc_id.encode())
        + _lp(_json_bytes(update_doc))
        + bytes([FLAG_RELAXED if relaxed else 0])
        + struct.pack(">Q", int(if_match))
    )


def encode_doc_del(collection: str, doc_id: str) -> bytes:
    return _lp(collection.encode()) + _lp(doc_id.encode())


def encode_query_by_id(collection: str, doc_id: str) -> bytes:
    return bytes([QUERY_BY_ID]) + _lp(collection.encode()) + _lp(doc_id.encode())


def _encode_bound(bound: Any) -> bytes:
    if bound is None:
        return b""
    values = bound if isinstance(bound, list) else [bound]
    return _json_bytes(values)


def encode_query_index_range(
    collection: str,
    index: str,
    *,
    lo: Any,
    hi: Any,
    page_size: int,
    cursor: bytes,
    include_bodies: bool = True,
) -> bytes:
    out = bytes([QUERY_INDEX_RANGE])
    out += _lp(collection.encode()) + _lp(index.encode())
    out += struct.pack(">I", page_size)
    out += _lp(_encode_bound(lo)) + _lp(_encode_bound(hi)) + _lp(cursor)
    # Append-only trailer: omit when true so legacy vectors match.
    if not include_bodies:
        out += b"\x00"
    return out


def _filter_bytes(filt: Optional[dict]) -> bytes:
    return b"" if not filt else _json_bytes(filt)


def encode_find(
    collection: str,
    filt: Optional[dict],
    sort: Optional[List[Tuple[str, bool]]],
    projection: Optional[Tuple[int, List[str]]],
    skip: int,
    limit: int,
    cursor: bytes,
) -> bytes:
    out = _lp(collection.encode()) + _lp(_filter_bytes(filt))
    sort = sort or []
    out += struct.pack(">I", len(sort))
    for field, ascending in sort:
        out += _lp(field.encode()) + bytes([1 if ascending else 0])
    if projection is None:
        out += bytes([PROJ_NONE])
    else:
        mode, fields = projection
        out += bytes([mode]) + struct.pack(">I", len(fields))
        for field in fields:
            out += _lp(field.encode())
    out += struct.pack(">II", skip, limit) + _lp(cursor)
    return out


def encode_update(
    collection: str,
    filt: dict,
    update_doc: dict,
    *,
    multi: bool,
    relaxed: bool,
    upsert: bool = False,
) -> bytes:
    flags = 0
    if relaxed:
        flags |= FLAG_RELAXED
    if upsert:
        flags |= FLAG_UPSERT
    out = (
        _lp(collection.encode())
        + _lp(_filter_bytes(filt))
        + _lp(_json_bytes(update_doc))
        + bytes([1 if multi else 0])
        + bytes([flags])
    )
    return out


def encode_delete(collection: str, filt: dict, *, multi: bool, relaxed: bool) -> bytes:
    return (
        _lp(collection.encode())
        + _lp(_filter_bytes(filt))
        + bytes([1 if multi else 0])
        + bytes([FLAG_RELAXED if relaxed else 0])
    )


def encode_count(collection: str, filt: Optional[dict]) -> bytes:
    return bytes([COUNT_MODE_COUNT]) + _lp(collection.encode()) + _lp(_filter_bytes(filt))


def encode_distinct(collection: str, field: str, filt: Optional[dict]) -> bytes:
    return (
        bytes([COUNT_MODE_DISTINCT])
        + _lp(collection.encode())
        + _lp(_filter_bytes(filt))
        + _lp(field.encode())
    )


def encode_aggregate(collection: str, pipeline: list) -> bytes:
    """Aggregate request: [collection lp][pipeline_json lp]."""
    return _lp(collection.encode()) + _lp(_json_bytes(pipeline))


def encode_watch(collection: str, resume_token: bytes = b"") -> bytes:
    """Watch request: [collection lp][resume_token lp]."""
    return _lp(collection.encode()) + _lp(resume_token)


def decode_watch_frame(buf: bytes) -> Tuple[str, bytes, Optional[int], Optional[bytes], Optional[bytes]]:
    """Decode one Watch stream frame.

    Returns ``(kind, resume_token, op, doc_id, body)`` where *kind* is
    ``"ack"``, ``"event"``, or ``"heartbeat"``; *op*/*doc_id*/*body* are set
    only for events.
    """
    if not buf:
        raise ValueError("empty watch frame")
    kind_byte = buf[0]
    off = 1
    (tlen,) = struct.unpack_from(">I", buf, off)
    off += 4
    resume_token = buf[off : off + tlen]
    off += tlen
    if kind_byte == WATCH_FRAME_ACK:
        return "ack", resume_token, None, None, None
    if kind_byte == WATCH_FRAME_HEARTBEAT:
        return "heartbeat", resume_token, None, None, None
    if kind_byte == WATCH_FRAME_EVENT:
        op = buf[off]
        off += 1
        (idlen,) = struct.unpack_from(">I", buf, off)
        off += 4
        doc_id = buf[off : off + idlen]
        off += idlen
        (blen,) = struct.unpack_from(">I", buf, off)
        off += 4
        body = buf[off : off + blen]
        return "event", resume_token, op, doc_id, body
    raise ValueError(f"unknown watch frame 0x{kind_byte:02x}")


def decode_aggregate_response(buf: bytes) -> List[bytes]:
    """Decode Aggregate response into raw row JSON byte strings."""
    off = 0
    (count,) = struct.unpack_from(">I", buf, off)
    off += 4
    rows: List[bytes] = []
    for _ in range(count):
        (n,) = struct.unpack_from(">I", buf, off)
        off += 4
        rows.append(buf[off : off + n])
        off += n
    return rows


def decode_page(buf: bytes) -> Tuple[List[Tuple[bytes, bytes]], bytes]:
    """Decode a query/find response page into `(rows, next_cursor)`.

    Each row is `(doc_id, body)`; an empty `next_cursor` means no more pages.
    """
    rows, cursor = decode_page_with_revision(buf, with_revision=False)
    return [(doc_id, body) for doc_id, body, _rev in rows], cursor


def decode_page_with_revision(
    buf: bytes, *, with_revision: bool = True
) -> Tuple[List[Tuple[bytes, bytes, Optional[int]]], bytes]:
    """Decode a FindRev page into `((doc_id, body, revision), next_cursor)`.

    When ``with_revision`` is False, ``revision`` is always ``None``.
    """
    off = 0
    (count,) = struct.unpack_from(">I", buf, off)
    off += 4
    rows: List[Tuple[bytes, bytes, Optional[int]]] = []
    for _ in range(count):
        (klen,) = struct.unpack_from(">I", buf, off)
        off += 4
        doc_id = buf[off : off + klen]
        off += klen
        (blen,) = struct.unpack_from(">I", buf, off)
        off += 4
        body = buf[off : off + blen]
        off += blen
        revision: Optional[int] = None
        if with_revision:
            (revision,) = struct.unpack_from(">Q", buf, off)
            off += 8
        rows.append((doc_id, body, revision))
    (clen,) = struct.unpack_from(">I", buf, off)
    off += 4
    cursor = buf[off : off + clen]
    return rows, cursor


def decode_doc_get_rev_response(buf: bytes) -> Tuple[bytes, int]:
    """Decode a DocGetRev response: `(body, revision)`."""
    (blen,) = struct.unpack_from(">I", buf, 0)
    body = buf[4 : 4 + blen]
    (revision,) = struct.unpack_from(">Q", buf, 4 + blen)
    return body, revision


def decode_begin_response(buf: bytes) -> Tuple[int, int]:
    """Decode Begin response: `(tx_id, snapshot_seq)`."""
    if len(buf) != 16:
        raise ValueError("Begin response must be 16 bytes")
    tx_id, snapshot_seq = struct.unpack(">QQ", buf)
    return tx_id, snapshot_seq


def decode_commit_response(buf: bytes) -> int:
    """Decode Commit response sequence number."""
    if len(buf) != 8:
        raise ValueError("Commit response must be 8 bytes")
    (seq,) = struct.unpack(">Q", buf)
    return seq


def decode_stage_ack(buf: bytes) -> Tuple[int, int]:
    """Decode stage ack: `(logical_ops, estimated_keys)`."""
    if len(buf) != 8:
        raise ValueError("stage ack must be 8 bytes")
    return struct.unpack(">II", buf)


def status_name(status: int) -> str:
    return STATUS_NAMES.get(status, f"0x{status:02x}")
